File: crates/core/src/records/subscribe/event.rs
use axum::response::sse::Event as SseEvent;
use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};
use std::sync::Arc;

use crate::records::RecordError;

type JsonObject = serde_json::value::Map<String, serde_json::Value>;

#[derive(Debug, Clone, Copy, Deserialize_repr, Serialize_repr, PartialEq)]
#[repr(i64)]
pub enum EventErrorStatus {
  /// Unknown or unspecified error.
  Unknown = 0,
  /// Access forbidden.
  Forbidden = 1,
  /// Server-side event-loss, e.g. a buffer ran out of capacity. This does not account for
  /// additional losses that may happen between the TrailBase server and the client. This
  /// needs to be determined client-side based on event `seq` numbers.
  Loss = 2,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
pub struct EventError {
  pub status: EventErrorStatus,
  pub message: Option<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum JsonEventPayload {
  Update { value: JsonObject },
  Insert { value: JsonObject },
  Delete { value: JsonObject },
  Error { value: EventError },
  Ping,
}

#[allow(unused)]
#[derive(Debug, Clone, Serialize)]
pub enum EventPayload {
  Update(Option<Box<serde_json::value::RawValue>>),
  Insert(Option<Box<serde_json::value::RawValue>>),
  Delete(Option<Box<serde_json::value::RawValue>>),
  Error(Option<Box<serde_json::value::RawValue>>),
  Ping,
}

impl PartialEq for EventPayload {
  fn eq(&self, other: &Self) -> bool {
    fn get(v: &Option<Box<serde_json::value::RawValue>>) -> Option<&str> {
      return v.as_ref().map(|v| v.get());
    }

    return match (self, other) {
      (Self::Update(lhs), Self::Update(rhs)) => get(lhs) == get(rhs),
      (Self::Insert(lhs), Self::Insert(rhs)) => get(lhs) == get(rhs),
      (Self::Delete(lhs), Self::Delete(rhs)) => get(lhs) == get(rhs),
      (Self::Error(lhs), Self::Error(rhs)) => get(lhs) == get(rhs),
      (Self::Ping, Self::Ping) => true,
      _ => false,
    };
  }
}

impl EventPayload {
  pub fn from(value: &JsonEventPayload) -> Self {
    return match value {
      JsonEventPayload::Update { value } => EventPayload::Update(
        serde_json::value::to_raw_value(&value)
          .map(|v| v.to_owned())
          .ok(),
      ),
      JsonEventPayload::Insert { value } => EventPayload::Insert(
        serde_json::value::to_raw_value(&value)
          .map(|v| v.to_owned())
          .ok(),
      ),
      JsonEventPayload::Delete { value } => EventPayload::Delete(
        serde_json::value::to_raw_value(&value)
          .map(|v| v.to_owned())
          .ok(),
      ),
      JsonEventPayload::Error { value } => EventPayload::Error(
        serde_json::value::to_raw_value(&value)
          .map(|v| v.to_owned())
          .ok(),
      ),
      JsonEventPayload::Ping => EventPayload::Ping,
    };
  }

  #[inline]
  pub fn into_sse_event(
    self: Arc<EventPayload>,
    seq: Option<i64>,
  ) -> Result<SseEvent, RecordError> {
    return match *self {
      Self::Ping => Ok(SseEvent::default().comment("ping")),
      _ => {
        let ev = ChangeEvent { event: self, seq };
        let s = serde_json::to_string(&ev).map_err(|err| RecordError::Internal(err.into()))?;
        Ok(SseEvent::default().data(&s))
      }
    };
  }

  #[cfg(feature = "ws")]
  #[inline]
  pub fn into_ws_event(self: Arc<EventPayload>) -> Result<axum::extract::ws::Message, RecordError> {
    return match *self {
      Self::Ping => Err(RecordError::Internal("not implemented".into())),
      _ => Ok(axum::extract::ws::Message::Text(
        serde_json::to_string(&*self)
          .map_err(|err| RecordError::Internal(err.into()))?
          .into(),
      )),
    };
  }
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct ChangeEvent {
  #[serde(flatten)]
  event: Arc<EventPayload>,
  // NOTE: Because unsigned isn't supported by Avro.
  #[serde(skip_serializing_if = "Option::is_none")]
  seq: Option<i64>,
}

#[cfg(test)]
#[derive(Debug, Clone, Deserialize, PartialEq)]
pub enum TestJsonEventPayload {
  Update(JsonObject),
  Insert(JsonObject),
  Delete(JsonObject),
  Error {
    status: EventErrorStatus,
    message: Option<String>,
  },
  Ping,
}

#[cfg(test)]
#[derive(Debug, Clone, Deserialize, PartialEq)]
pub struct TestChangeEvent {
  #[serde(flatten)]
  pub event: TestJsonEventPayload,
  pub seq: Option<i64>,
}

#[cfg(test)]
mod tests {
  use super::*;
  use serde_json::json;

  #[test]
  fn serialization_sse_event_test() {
    {
      let event = ChangeEvent {
        event: Arc::new(EventPayload::from(&JsonEventPayload::Delete {
          value: JsonObject::from_iter([("foo".to_string(), json!(4))]),
        })),
        seq: Some(4),
      };

      let value = serde_json::to_value(&event).unwrap();
      assert_eq!(
        serde_json::json!({
            "Delete": {
                "foo": 4,
            },
            "seq": 4,
        }),
        value
      );
    }

    {
      let event = ChangeEvent {
        event: Arc::new(EventPayload::from(&JsonEventPayload::Error {
          value: EventError {
            status: EventErrorStatus::Loss,
            message: Some("test".to_string()),
          },
        })),
        seq: Some(4),
      };

      let expected = serde_json::json!({
          "Error": {
              "status": EventErrorStatus::Loss,
              "message": "test",
          },
          "seq": 4,
      });

      assert_eq!(expected, serde_json::to_value(&event).unwrap());
    }

    {
      let json = r#"
            {
              "Error": {
                "status": 1,
                "message": "test"
              },
              "seq": 3
            }
        "#;

      let test_event = serde_json::from_str::<TestChangeEvent>(json).unwrap();
      assert_eq!(Some(3), test_event.seq);
      assert_eq!(
        TestJsonEventPayload::Error {
          status: EventErrorStatus::Forbidden,
          message: Some("test".to_string()),
        },
        test_event.event
      );
    }
  }
}

File: client/dotnet/trailbase/Event.cs
using System.Text.Json.Nodes;

namespace TrailBase;

/// <summary>Realtime event for change subscriptions.</summary>
public abstract class Event {
  /// <summary>Get associated record value as JSON object.</summary>
  public abstract JsonNode? Value { get; }
  /// <summary>Get associated error message.</summary>
  public long? Seq { get; }

  /// <summary>InsertEvent constructor.</summary>
  public Event(long? seq) {
    this.Seq = seq;
  }

  /// <summary>Parses JSON into an Event.</summary>
  public static Event Parse(string message) {
    var obj = (JsonObject?)JsonNode.Parse(message);
    if (obj != null) {
      var seq = obj["seq"]?.GetValue<long>();

      var insert = obj["Insert"];
      if (insert != null) {
        return new InsertEvent(seq, insert);
      }

      var update = obj["Update"];
      if (update != null) {
        return new UpdateEvent(seq, update);
      }

      var delete = obj["Delete"];
      if (delete != null) {
        return new DeleteEvent(seq, delete);
      }

      var error = obj["Error"];
      if (error != null) {
        return new ErrorEvent(seq, (ErrorEvent.ErrorStatus)(error["status"]?.GetValue<int>() ?? 0), error["message"]?.GetValue<string>());
      }
    }

    throw new Exception($"Failed to parse {message}");
  }
}

/// <summary>Record insertion event.</summary>
public class InsertEvent : Event {
  /// <summary>Get associated record value as JSON object.</summary>
  public override JsonNode? Value { get; }

  /// <summary>InsertEvent constructor.</summary>
  public InsertEvent(long? seq, JsonNode? value) : base(seq) {
    this.Value = value;
  }

  /// <summary>Serialize InsertEvent.</summary>
  public override string ToString() => $"InsertEvent({Value})";
}

/// <summary>Record update event.</summary>
public class UpdateEvent : Event {
  /// <summary>Get associated record value as JSON object.</summary>
  public override JsonNode? Value { get; }

  /// <summary>UpdateEvent constructor.</summary>
  public UpdateEvent(long? seq, JsonNode? value) : base(seq) {
    this.Value = value;
  }

  /// <summary>Serialize UpdateEvent.</summary>
  public override string ToString() => $"UpdateEvent({Value})";
}

/// <summary>Record deletion event.</summary>
public class DeleteEvent : Event {
  /// <summary>Get associated record value as JSON object.</summary>
  public override JsonNode? Value { get; }

  /// <summary>DeleteEvent constructor.</summary>
  public DeleteEvent(long? seq, JsonNode? value) : base(seq) {
    this.Value = value;
  }

  /// <summary>Serialize DeleteEvent.</summary>
  public override string ToString() => $"DeleteEvent({Value})";
}

/// <summary>Error event.</summary>
public class ErrorEvent : Event {
  /// <summary>Programmatic error status.</summary>
  public enum ErrorStatus : int {
    /// Unknown error status.
    Unknown = 0,
    /// Forbidden, i.e. ACL violation.
    Forbidden = 1,
    /// Server-side event loss. Independently events can get lost between TrailBase and the client.
    Loss = 2,
  }

  /// <summary>Get associated record value as JSON object.</summary>
  public override JsonNode? Value { get { return null; } }

  /// <summary>Programmatic status of this error event.</summary>
  public ErrorStatus Status { get; }
  /// <summary>Get associated error message.</summary>
  public string? Message { get; }

  /// <summary>ErrorEvent constructor.</summary>
  public ErrorEvent(long? seq, ErrorStatus status, string? message) : base(seq) {
    this.Status = status;
    this.Message = message;
  }

  /// <summary>Serialize ErrorEvent.</summary>
  public override string ToString() => $"ErrorEvent({Status}, {Message})";
}
